src/LSST/control/ControlMain.cpp

```diff
@@ -57,16 +66,161 @@ ControlMain::Ptr ControlMain::getPtr() {
+ControlMain::ControlMain() {}
+
+ControlMain::~ControlMain() {}
+
+void ControlMain::run(int argc, const char* argv[]) {
+        thread t(&ControlMain::_cMain, this, argc, argv);
+        _mainThrd = move(t);
+}
+
+
+void signalHandler(int sig) {
+    if (sig == SIGPIPE) {
+        LWARN("Ignoring SIGPIPE sig=", sig);
+        return;
     }
-    return *_thisPtr;
+    LCRITICAL("Unhandled signal caught sig=", sig);
+    exit(sig);
 }
 
-ControlMain::ControlMain() {}
+void ControlMain::_cMain(int argc, const char* argv[]) {
+    LINFO("starting main");
+    util::Log& log = util::Log::getLog();
 
-ControlMain::~ControlMain() {}
+    // Get the configuration
+    system::Config& sysCfg = system::Config::get();
+
+    // Setup global items.
+    system::Globals::setup(sysCfg);
+
+    // Setup a simple signal handler to handle when clients closing connection results in SIGPIPE.
+    signal(SIGPIPE, signalHandler);
+
+    // Create the control system
+    _simCore.reset(new LSST::m2cellcpp::simulator::SimCore());
+    _simCore->start();
+    faultmgr::FaultMgr::setup();
+    control::FpgaIo::setup(_simCore);
+    control::MotionEngine::setup();
+    control::Context::setup();
+
+    // Register the PowerSystem with FpgaIo so that it gets updates.
+    auto powerSys = control::Context::get()->model.getPowerSystem();
+    control::FpgaIo::get().registerPowerSys(powerSys);
+
+    // Setup the control system
+    auto context = control::Context::get();
+    context->model.ctrlSetup();
+
+    // At this point, MotionCtrl and FpgaCtrl should be configured. Start the control loops
+    context->model.ctrlStart();
+
+    // At this point, the LabView code seems to want to put the system into
+    // ReadyIdle. We're NOT doing that. The system is going into StandbyState
+    // until it gets an explicit command to do something else.
+
+    // Start the telemetry server
+    // FUTURE: Telemetry for the GUI requires the ComControlServer->ComConnection::welcomeMsg to
+    //         work properly. Those elements should be in the telemetry so there's no need for a
+    //         ComControlServer connection for Telemetry. This has to wait until the existing
+    //         controller is no longer used.
+    LINFO("Starting Telemetry Server");
+    // FUTURE: get the correct entries into `system::Config`.
+    int const telemPort = 50001;
+    auto servTelemetryMap = system::TelemetryMap::Ptr(new system::TelemetryMap());
+    auto telemetryServ = system::TelemetryCom::create(servTelemetryMap, telemPort);
+
+    telemetryServ->startServer();
+    if (!telemetryServ->waitForServerRunning(5)) {
+        LCRITICAL("Telemetry server failed to start.");
+        exit(-1);  // change to powerdown and exit
+    }
+
+    // Wait for MotionCtrl to be ready
+    context->model.waitForCtrlReady();
+
+    // Start a ComControlServer
+    LDEBUG("ComControlServer starting...");
+    system::IoContextPtr ioContext = make_shared<boost::asio::io_context>();
```

Comment:
This may not be a good idea to hide the original type as it makes it more difficult to understand what's behind such indirection (`system::IoContextPtr` instead of `std_shared<boost::asio::io_context>`).

And by the way, do you really need a pointer instead of the instance here (in the `main`)? You may just have the data member:
```
class ControlMain {
private:
    boost::asio::io_context _io_context;
};
```
An assumption is that the life expectancy of `ControlMain` is loner than anything within this application.

Reply:
ControlMain runs within a thread, so there could be lifetime issues and this change would probably break all of the existing unit tests. It may still be a good idea to change this and make a boost::asio refactoring ticket.

`system::IoContextPtr` is used a lot in the `system` namespace. It's pretty clear what it is.